// ngx-plaid-link/esm5/lib/interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PlaidSuccessMetadata() { }
/** @type {?} */
PlaidSuccessMetadata.prototype.link_session_id;
/** @type {?} */
PlaidSuccessMetadata.prototype.institution;
/** @type {?} */
PlaidSuccessMetadata.prototype.account;
/**
 * @record
 */
export function PlaidOnSuccessArgs() { }
/** @type {?} */
PlaidOnSuccessArgs.prototype.token;
/** @type {?} */
PlaidOnSuccessArgs.prototype.metadata;
/**
 * @record
 */
export function PlaidInstitutionObject() { }
/** @type {?} */
PlaidInstitutionObject.prototype.name;
/** @type {?} */
PlaidInstitutionObject.prototype.institution_id;
/**
 * @record
 */
export function PlaidAccountObject() { }
/** @type {?} */
PlaidAccountObject.prototype.id;
/** @type {?} */
PlaidAccountObject.prototype.name;
/** @type {?} */
PlaidAccountObject.prototype.mask;
/** @type {?} */
PlaidAccountObject.prototype.type;
/** @type {?} */
PlaidAccountObject.prototype.subtype;
/**
 * @record
 */
export function PlaidErrorObject() { }
/** @type {?} */
PlaidErrorObject.prototype.display_message;
/** @type {?} */
PlaidErrorObject.prototype.error_code;
/** @type {?} */
PlaidErrorObject.prototype.error_message;
/** @type {?} */
PlaidErrorObject.prototype.error_type;
/**
 * @record
 */
export function PlaidErrorMetadata() { }
/** @type {?} */
PlaidErrorMetadata.prototype.link_session_id;
/** @type {?} */
PlaidErrorMetadata.prototype.institution;
/** @type {?} */
PlaidErrorMetadata.prototype.status;
/**
 * @record
 */
export function PlaidOnExitArgs() { }
/** @type {?} */
PlaidOnExitArgs.prototype.error;
/** @type {?} */
PlaidOnExitArgs.prototype.metadata;
/**
 * @record
 */
export function PlaidOnEventArgs() { }
/** @type {?} */
PlaidOnEventArgs.prototype.eventName;
/** @type {?} */
PlaidOnEventArgs.prototype.metadata;
/**
 * @record
 */
export function PlaidEventMetadata() { }
/** @type {?} */
PlaidEventMetadata.prototype.error_code;
/** @type {?} */
PlaidEventMetadata.prototype.error_message;
/** @type {?} */
PlaidEventMetadata.prototype.error_type;
/** @type {?} */
PlaidEventMetadata.prototype.exit_status;
/** @type {?} */
PlaidEventMetadata.prototype.institution_id;
/** @type {?} */
PlaidEventMetadata.prototype.institution_name;
/** @type {?} */
PlaidEventMetadata.prototype.institution_search_query;
/** @type {?} */
PlaidEventMetadata.prototype.request_id;
/** @type {?} */
PlaidEventMetadata.prototype.link_session_id;
/** @type {?} */
PlaidEventMetadata.prototype.mfa_type;
/** @type {?} */
PlaidEventMetadata.prototype.view_name;
/** @type {?} */
PlaidEventMetadata.prototype.timestamp;
/**
 * @record
 */
export function PlaidConfig() { }
/** @type {?|undefined} */
PlaidConfig.prototype.apiVersion;
/** @type {?|undefined} */
PlaidConfig.prototype.clientName;
/** @type {?} */
PlaidConfig.prototype.env;
/** @type {?} */
PlaidConfig.prototype.key;
/** @type {?|undefined} */
PlaidConfig.prototype.onLoad;
/** @type {?} */
PlaidConfig.prototype.onSuccess;
/** @type {?} */
PlaidConfig.prototype.onExit;
/** @type {?|undefined} */
PlaidConfig.prototype.onEvent;
/** @type {?} */
PlaidConfig.prototype.product;
/** @type {?|undefined} */
PlaidConfig.prototype.token;
/** @type {?|undefined} */
PlaidConfig.prototype.webhook;
